petastorm/pytorch.py

```diff
@@ -122,60 +122,84 @@ def __init__(self, reader, batch_size=1, collate_fn=decimal_friendly_collate,
+
+        # _batch_acc accumulates samples for a single batch.
+        self._batch_acc = []
         if shuffling_queue_capacity > 0:
-            # We always add and remove one sample from the shuffling buffer in the steady state. min_after_dequeue
-            # is important when enqueue and dequeue rates are not the same.
-            # By setting `min_after_dequeue = shuffling_queue_capacity - 1` we will just keep the buffer full.
+            # We can not know what is the reasonable number to use for the extra capacity, so we set a huge number
+            # and give up on the unbound growth protection mechanism.
             min_after_dequeue = shuffling_queue_capacity - 1
             self._shuffling_buffer = RandomShufflingBuffer(shuffling_queue_capacity,
-                                                           min_after_retrieve=min_after_dequeue, extra_capacity=0)
+                                                           min_after_retrieve=min_after_dequeue,
+                                                           extra_capacity=100000000)
         else:
             self._shuffling_buffer = NoopShufflingBuffer()
 
     def __iter__(self):
         """
         The Data Loader iterator stops the for-loop when reader runs out of samples.
         """
-        batch = []
+        # As we iterate over incoming samples, we are going to store them in `self._batch_acc`, until we have a batch of
+        # the requested batch_size ready.
+
+        keys = None
+
         for row in self.reader:
             # Default collate does not work nicely on namedtuples and treat them as lists
             # Using dict will result in the yielded structures being dicts as well
             row_as_dict = row._asdict()
 
+            keys = row_as_dict.keys()
+
             # Promote some types that are incompatible with pytorch to be pytorch friendly.
             _sanitize_pytorch_types(row_as_dict)
 
             # Add rows to shuffling buffer
-            self._shuffling_buffer.add_many([row_as_dict])
-
-            # If the shuffling buffer is ready, pull the resampled rows out and try add to our accumulated batch
-            while self._shuffling_buffer.can_retrieve():
-                post_shuffled_row = self._shuffling_buffer.retrieve()
-                batch.append(post_shuffled_row)
-
-                # Batch is ready? Collate and emmit
-                if len(batch) == self.batch_size:
-                    yield self.collate_fn(batch)
-                    batch = []
-
-        # Once reader can not emit new rows, we might still have a bunch of rows waiting in the shuffling buffer.
-        # Telling shuffling buffer thatt we are finished allows to deplete the buffer completely, regardless its
+            if not self.reader.is_batched_reader:
+                self._shuffling_buffer.add_many([row_as_dict])
+            else:
+                # Transposition:
+                #   row_as_dict:        {'a': [1,2,3], 'b':[4,5,6]}
```

Comment:
Is it always true that a and b will have the same length of data

Reply:
Good point. It should be so for scalar fields. Need to check if it works properly in case of lists.